base/abstractdict.jl

```diff
@@ -335,6 +335,10 @@ function _typeddict(d::AbstractDict, others::AbstractDict...)
     Dict{K,V}(d)
 end
 
+reduce(m::typeof(merge), items::AbstractVector) = _merge!(mapreduce(typeof, typejoin, items), items)
+_toany(T) = isconcretetype(T) ? T : Any
```

Comment:
This function isn't used in this file, better move it to the other one. But is it really needed? The test which mentions `Any` vs. `Real` doesn't sound correct since one of the dicts uses `Symbol` keys.

Actually, it seems what is needed is not to use `K` and `V` at all (i.e. do `::Type{<:Dict}`), but instead compute them with `mapreduce(keytype, promote_type, items)` and `mapreduce(valuetype, promote_type, items)`.

Reply:
Agreed it should be moved. And, I agree it is not tested properly. If I recall, the performance with, say `Real` rather than `Any` was much worse. But, that can be optimized later, and may be better or worse by now with changes to v0.7 in the last month. The main reason is that I think `Any` is better than a more specific abstract type. There must be some discussion on this kind of thing. It is not mentioned explicitly in the [style guide](https://docs.julialang.org/en/latest/manual/style-guide/). But it  is similar in spirit to the admonition against weird `Union` types.

What do you think about moving all the new code to `dict.jl` ? The new lines in `abstractdict.jl` don't mention `AbstractDict` or subtypes thereof at all.

> Actually, it seems what is needed is not to use K and V at all (i.e. do ::Type{<:Dict}), but instead compute them with 

I think you are suggesting to use `mapreduce` to compute each of the key and value types rather than the type of `Dict` at one go ?  This sounds ok. But, that still may result in `Dict{A,Real}` etc. It would be nice to get some feedback or testing in the wild before it is frozen in 1.0.

Reply:
> Agreed it should be moved. And, I agree it is not tested properly. If I recall, the performance with, say Real rather than Any was much worse. But, that can be optimized later, and may be better or worse by now with changes to v0.7 in the last month. The main reason is that I think Any is better than a more specific abstract type. There must be some discussion on this kind of thing. It is not mentioned explicitly in the style guide. But it is similar in spirit to the admonition against weird Union types.

I don't think we need to think about what's the best type here: it must be the same as what `merge(::Dict, ::Dict)` gives, that's all.



> What do you think about moving all the new code to dict.jl ? The new lines in abstractdict.jl don't mention AbstractDict or subtypes thereof at all.

Yeah, on the other hand `merge` is defined here to it makes sense to define `reduce(::typeof(merge), ...)` here too.

> I think you are suggesting to use mapreduce to compute each of the key and value types rather than the type of Dict at one go ? This sounds ok. But, that still may result in Dict{A,Real} etc. It would be nice to get some feedback or testing in the wild before it is frozen in 1.0.

Yes, but AFAIK that's what `merge` does already, so that what we should do. If we want to change that, better file a separate issue.